pkg/stores/proxyalpha/proxy_store.go

```diff
@@ -0,0 +1,678 @@
+type WarningBuffer []types.Warning
+
+// HandleWarningHeader takes the components of a kubernetes warning header and stores them
+func (w *WarningBuffer) HandleWarningHeader(code int, agent string, text string) {
+	*w = append(*w, types.Warning{
+		Code:  code,
+		Agent: agent,
+		Text:  text,
+	})
+}
+
+// RelationshipNotifier is an interface for handling wrangler summary.Relationship events.
+type RelationshipNotifier interface {
+	OnInboundRelationshipChange(ctx context.Context, schema *types.APISchema, namespace string) <-chan *summary.Relationship
+}
+
+type Store struct {
+	clientGetter   ClientGetter
+	notifier       RelationshipNotifier
+	cacheFactory   CacheFactory
+	cfInitializer  CacheFactoryInitializer
+	namespaceCache Cache
+	lock           sync.Mutex
+	columnSetter   SchemaColumnSetter
+}
+
+type CacheFactoryInitializer func() (CacheFactory, error)
+
+type CacheFactory interface {
+	CacheFor(fields [][]string, client dynamic.ResourceInterface, gvk schema.GroupVersionKind, namespaced bool) (factory.Cache, error)
+	Reset() error
+}
+
+// NewProxyStore returns a Store implemented directly on top of kubernetes.
+func NewProxyStore(c SchemaColumnSetter, clientGetter ClientGetter, notifier RelationshipNotifier, factory CacheFactory) (*Store, error) {
+	store := &Store{
+		clientGetter: clientGetter,
+		notifier:     notifier,
+		columnSetter: c,
+	}
+
+	if factory == nil {
+		var err error
+		factory, err = defaultInitializeCacheFactory()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	store.cacheFactory = factory
+	if err := store.initializeNamespaceCache(); err != nil {
+		logrus.Infof("failed to warm up namespace informer for proxy store in steve, will try again on next ns request")
+	}
+	return store, nil
+}
+
+// Reset locks the store, resets the underlying cache factory, and warm the namespace cache.
+func (s *Store) Reset() error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	if err := s.cacheFactory.Reset(); err != nil {
+		return err
+	}
+
+	if err := s.initializeNamespaceCache(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func defaultInitializeCacheFactory() (CacheFactory, error) {
+	informerFactory, err := factory.NewCacheFactory()
+	if err != nil {
+		return nil, err
+	}
+	return informerFactory, nil
+}
+
+func (s *Store) initializeNamespaceCache() error {
+	buffer := WarningBuffer{}
+	nsSchema := baseNSSchema
+	if err := s.columnSetter.SetColumns(context.TODO(), &nsSchema); err != nil {
+		return fmt.Errorf("failed to set columns for proxy stores namespace informer: %w", err)
+	}
+	client, err := s.clientGetter.TableAdminClient(nil, &nsSchema, "", &buffer)
+	if err != nil {
+		return err
+	}
+
+	fields := getFieldsFromSchema(&nsSchema)
+	fields = append(fields, getFieldForGVK(attributes.GVK(&nsSchema))...)
+	nsInformer, err := s.cacheFactory.CacheFor(fields, &tablelistconvert.Client{ResourceInterface: client}, attributes.GVK(&nsSchema), false)
+	if err != nil {
+		return err
+	}
+
+	s.namespaceCache = nsInformer
```

Comment:
Stopping the informers is done by the factory via the stop channel. This is an unexported function only called when we have either reset the factory or are creating a new one. If a namespace informer is set then it will have been stopped by Reset(), or it is the exact same informer as the one being assigned.